akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,377 @@
+  private val pendingInvocations = new java.util.LinkedList[PendingHandlerInvocation]() // we only append / isEmpty / get(0) on it
+  private var eventBatch: List[PersistentEnvelope] = Nil
+
+  private var journalBatch = Vector.empty[PersistentEnvelope]
+  private var writeInProgress = false
+
+  // ----------
+
+  private def lastSequenceNr: Long = sequenceNr
+  private def snapshotSequenceNr: Long = lastSequenceNr
+
+  private def updateLastSequenceNr(persistent: PersistentRepr): Unit =
+    if (persistent.sequenceNr > lastSequenceNr) sequenceNr = persistent.sequenceNr
+  private def nextSequenceNr(): Long = {
+    sequenceNr += 1L
+    sequenceNr
+  }
+  // ----------
+
+  private def onSnapshotterResponse(response: SnapshotProtocol.Response): Behavior[Any] = {
+    response match {
+      case SaveSnapshotSuccess(meta)     ⇒ log.info("Save snapshot successful: " + meta)
+      case SaveSnapshotFailure(meta, ex) ⇒ log.error(ex, "Save snapshot failed! " + meta) // FIXME no fail? no callback?
+    }
+    same
+  }
+
+  // ----------
+
+  trait EventsourcedRunningPhase {
+    def name: String
+    def onCommand(c: Command): Behavior[Any]
+    def onJournalResponse(response: JournalProtocol.Response): Behavior[Any]
+  }
+
+  object HandlingCommands extends EventsourcedRunningPhase {
+    def name = "HandlingCommands"
+
+    final override def onCommand(command: Command): Behavior[Any] = {
+      val effect = commandHandler(commandContext, state, command)
+      applyEffects(command, effect)
+    }
+    final override def onJournalResponse(response: Response): Behavior[Any] = {
+      // should not happen, what would it reply?
+      throw new RuntimeException("Received message which should not happen in Running state!")
+    }
+  }
+
+  object PersistingEvents extends EventsourcedRunningPhase {
+    def name = "PersistingEvents"
+    final override def onCommand(c: Command): Behavior[Any] = {
+      log.info(s"PERSISTING EVENTS, command, STASH: ${c}")
+      stash(c, same)
+    }
+    final override def onJournalResponse(response: Response): Behavior[Any] = {
+      log.info("RESPONSE == " + response)
+      response match {
+        case WriteMessageSuccess(p, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case we ignore the call to the handler
+          if (id == writerIdentity.instanceId) {
+            updateLastSequenceNr(p)
+            popApplyHandler(p.payload)
+            onWriteMessageComplete()
+          } else same
+
+        case WriteMessageRejected(p, cause, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case the handler has already been discarded
+          if (id == writerIdentity.instanceId) {
+            updateLastSequenceNr(p)
+            onWriteMessageComplete() // FIXME no sure if we should call it here
```

Comment:
not sure but something could be wrong here, `onWriteMessageComplete` is calling `tryUnstash`